aten/src/ATen/native/cuda/LinearAlgebra.cu

```diff
@@ -492,8 +495,70 @@ void addr_kernel_cuda(TensorIterator &iter, Scalar beta, Scalar alpha) {
   });
 }
 
+// This reduction accumulates results as the type `acc_t`. By default, when
+// `scalar_t` is complex, `acc_t` is the downgraded real number type.
+// Otherwise, `acc_t` and `scalar_t` are the same type.
+template <typename scalar_t, typename acc_t=typename scalar_value_type<scalar_t>::type, typename out_t=typename scalar_value_type<scalar_t>::type>
+void linalg_vector_norm_kernel_cuda_impl(TensorIterator& iter, Scalar val) {
+  double ord_val;
+  if (val.isIntegral(/*includeBool=*/false)) {
+     ord_val = val.to<int64_t>();
+  } else if (val.isFloatingPoint()) {
+     ord_val = val.to<double>();
+  } else {
+     TORCH_CHECK(false, "linalg_vector_norm_kernel_cuda_impl expects norm to be integer or float");
+  }
+  if (iter.numel() == 0) {
+    iter.output().fill_((ord_val < 0) ? INFINITY : 0);
+    return;
+  }
+  if (ord_val == 0) {
+    gpu_reduce_kernel<scalar_t, out_t>(iter, NormZeroOps<scalar_t, acc_t>(), 0);
+  } else if (ord_val == 1) {
+    gpu_reduce_kernel<scalar_t, out_t>(iter, NormOneOps<scalar_t, acc_t>(), 0);
+  } else if (ord_val == 2) {
+    gpu_reduce_kernel<scalar_t, out_t>(iter, NormTwoOps<scalar_t, acc_t>(), 0);
+  } else if (ord_val == INFINITY) {
+    gpu_reduce_kernel<scalar_t, out_t>(iter, AbsMaxOps<scalar_t, acc_t>(), 0);
+  } else if (ord_val == -INFINITY) {
+    gpu_reduce_kernel<scalar_t, out_t>(iter, AbsMinOps<scalar_t, acc_t>(), std::numeric_limits<acc_t>::infinity());
+  } else {
+    gpu_reduce_kernel<scalar_t, out_t>(iter, NormOps<scalar_t, acc_t>{ static_cast<acc_t>(ord_val) }, 0);
+  }
+  // For complex outputs, the above kernels do not touch the imaginary values,
+  // so we must zero them out
+  if (isComplexType(iter.output().scalar_type())) {
+    at::imag(iter.output()).zero_();
+  }
+}
+
+static void linalg_vector_norm_kernel_cuda(TensorIterator& iter, Scalar ord) {
+#if defined(__HIPCC__)
+  if (iter.input_dtype() == kHalf) {
+    return linalg_vector_norm_kernel_cuda_impl<at::Half, float>(iter, ord);
+  } else if (iter.dtype(1) == kHalf && iter.input_dtype() == kFloat) {
```

Comment:
How is `iter.dtype(1)` distinct from `iter.input_dtype()`? A comment here identifying the inputs/outputs and their positions would help, too.

Reply:
In this case, `iter.dtype(1)` is the output's dtype. I'll change it to `iter.output_dtype()`, since that's clearer

Reply:
Wait I thought that outputs had to appear before inputs in TensorIterator?

Reply:
Wait, how can `iter.dtype(1)` be input dtype? TI does not work this way. `iter.output_dtype()` translates into `iter.dtype(1)` for an op with 1 output.

Reply:
Unless I'm misunderstanding something, I think Mike is right--outputs are placed in front of inputs in `TensorIterator::operands_`. When calling `TensorIterator::input(int arg)`, `arg` is offset by the number of outputs: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/TensorIterator.h#L219

So `iter.dtype(1)` is equivalent to `iter.input_dtype()`. That can also be verified by looking at the implementation for `input_dtype()` and `dtype()`: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/TensorIterator.h#L198 and https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/TensorIterator.h#L193

This means that the above `iter.dtype(1) == kHalf && iter.input_dtype() == kFloat` condition is always false. I copied this code from `norm_kernel_cuda`, assuming that it was correct: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/ReduceNormKernel.cu#L42

But it turns out that this code is incorrect because of a commit I made. The conditional used to be `iter.dtype(1) == kHalf && iter.dtype() == kFloat`. This is almost definitely where I introduced the change that broke float16 performance for CUDA.

I'll fix this in this PR for `vector_norm`. Should we also fix it for `torch.norm` in this PR, or would it be best to do that in another PR?

Also, as side note, it turns out `TensorIterator::output_dtype()` does not exist, so I assume it's probably best to use `iter.output().scalar_type()` for readability.

Reply:
The change that broke half performance was the change to `make_reduction` invocation: https://github.com/pytorch/pytorch/commit/54f0556ee453d1c4a9c6756bfcdce1e0d62867a8#diff-105d96b43a1224f639b222181fc67530f4edea19eacd33db8c57fe6273e0303cL678, `make_reduction` has to be called with out_dtype only to do the right thing (this is very fragile and should be fixed really). `make_reduction` itself, when called wrong, is doing the case, so after it nothing can be done.

Reply:
I'm happy to fix the `make_reduction` call, but could you please give a bit more detail? I don't understand why using `out_dtype` is incorrect or how it should be fixed.

Maybe an alternative to `out_dtype` would be to create an empty result tensor directly (only if it's not defined already) and pass that into `make_reduction`? Something like this:

```
if (!result.defined()) {
  result = at::empty({0}, self.options().dtype(out_dtype));
}
auto iter = make_reduction("vector_norm", result, self_, dim, keepdim, in_dtype);
```

Is that what you'd prefer @ngimel, or something else?

Reply:
I tried my own suggestion from above, and it doesn't work. If result is defined and its dtype is not the same as self_, then I get an error. I also get an error if result is not defined and the input is complex:

```
>>> a = torch.randn(10, dtype=torch.cfloat)
>>> result_out = torch.empty((0), dtype=torch.float)
>>> torch.linalg.vector_norm(a, 0, out=result_out)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: vector_norm: provided dtype must match dtype of result. Got Float and ComplexFloat.
>>> torch.linalg.vector_norm(a, 0)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: vector_norm: provided dtype must match dtype of result. Got Float and ComplexFloat.
```

I'm not sure how else we could avoid the `out_dtype` argument for `make_reduction`

Reply:
Hey @kurtamohler, @ngimel and @mruberry here. Let's talk about two things:

- First, let's fix that bug in this PR here and for torch.norm and ensure the code path is exercised properly whether that's with a test (if possible) or just manually by verifying a print statement is triggered, for example.
- Second, let's talk about make_reduction(). This is the overload that would need to be called for special half handling:

https://github.com/pytorch/pytorch/blob/c9d37675b2a44794d04c33bdaba0c40c7efabece/aten/src/ATen/native/ReduceOpsUtils.h#L196

And we would like to restore that functionality, but NOT in this PR.

Reply:
I believe I've fixed this for `vector_norm`. But only two of these four conditions are being tested (verified with a print statement)--both of the cases that check if the output is a half precision type. The other two conditions cannot be tested at the moment, because they rely on the input and output types being different (half and float, respectively). Since this PR is throwing an error if the result tensor's type does not match the `dtype` argument (if given) or the value type of the input, then the TensorIterator's output and input dtypes will always match